test/testdata/infer/private_methods.rb

```diff
@@ -0,0 +1,66 @@
+# typed: true
+
+class Test
+  def using_symbol
+    1
+  end
+  private :using_symbol
+
+  private def using_symbol_returned_by_def
+    2
+  end
+
+  private def splat_call(*args)
+    puts args
+  end
+
+  private def splat_and_block_call(foo, bar, &blk)
+  end
+
+  private def block_call(&blk)
+  end
+
+  def calling_private
+    using_symbol
+    self.using_symbol
+    assigned_self = self
+    assigned_self.using_symbol # error: Non-private call to private method `Test#using_symbol`
+  end
+
+  private
+  def subsequent_visibility
+    3
+  end
+end
+
+class TestChild < Test
+  def calling_private_in_parent
+    using_symbol
+    self.using_symbol
+    assigned_self = self
+    assigned_self.using_symbol # error: Non-private call to private method `Test#using_symbol`
+  end
+end
+
+Test.new.using_symbol # error: Non-private call to private method `Test#using_symbol`
+Test.new.using_symbol_returned_by_def # error: Non-private call to private method `Test#using_symbol_returned_by_def`
+Test.new.calling_private
+Test.new.using_symbol { 123 } # error: Non-private call to private method `Test#using_symbol`
+Test.new.block_call # error: Non-private call to private method `Test#block_call`
+Test.new.block_call { 123 } # error: Non-private call to private method `Test#block_call`
+Test.new.block_call(&:foo) # error: Non-private call to private method `Test#block_call`
+
+TestChild.new.using_symbol # error: Non-private call to private method `Test#using_symbol`
+
+T.unsafe(Test.new).using_symbol
+
+# TODO: The following methods should also error out. The do not currently do that since the the IR
+# of these calls are through the magic calls: splat-call, splat-and-block-call and block-call.
+# They don't contain the visibility information of the original method.
+Test.new.splat_call(*T.unsafe(nil))
+Test.new.splat_and_block_call(*[1, 'a'], &nil)
+Test.new.block_call(&nil)
```

Comment:
Can you leave the TODOs in the code where the changes would be required?

Also, I don't think it would be hard to support these off the bat, would it? Can you similarly pass along the flags when synthesizing a `<Magic>.<callWithSplat>` method call?

Reply:
I can take a quick stab before declaring to wait. Figured the probability of hitting a method marked both private and being a splat and/or block method was on the lower side, so it could wait.

Reply:
Took another look at this. I do pass on the flags of `isPrivateOk` when for `MK::Send`, so that part is covered. What's missing is to know if the method is private or not.
Please correct me if I misunderstood the built-in magic methods like `call-with-block`, but as far as I understand there only exist a single method for `call-with-block`, no matter how many times or on which methods it's used:

https://github.com/sorbet/sorbet/blob/7b333e9c50a020924503b0abf8712ee118f8bae1/core/GlobalState.cc#L519-L530

So I don't think I can set the private method flag (`setMethodPrivate`) on that without doing so across all usages of it?

Perhaps we could split the `callWithBlock` magic method into two methods: `callWithBlockPublic` & `callWithBlockPrivate`. But that doesn't feel right design-wise to me.

Reply:
You're right, that's not the right approach. I thought it would be as simple as setting the flags on the constructed send in Desugar.cc but that wasn't enough. I'm fine not blocking on that to land this.